java/ql/src/semmle/code/java/dataflow/internal/DataFlowUtil.qll

```diff
@@ -403,3 +404,27 @@ Node getInstanceArgument(Call call) {
   explicitInstanceArgument(call, result.asExpr()) or
   implicitInstanceArgument(call, result.(ImplicitInstanceAccess).getInstanceAccess())
 }
+
+/**
+ * A guard that validates some expression.
+ *
+ * To use this in a configuration, extend the class and provide a
+ * characteristic predicate precisely specifying the guard, and override
+ * `checks` to specify what is being validated and in which branch.
+ *
+ * It is important that all extending classes in scope are disjoint.
```

Comment:
@xiemaisi Did you know about this issue? It should affect JavaScript too, but we couldn't find a non-contrived example where it would make a difference.

Reply:
The worst we could come up with was something like this:
```
var.regexMatches(safePattern)
var.regexMatches(patternToFindBadStuff)
```
If the two `BarrierGuard`s describing those two case were sufficiently loosely described in the charpreds, such that there might exist regex patterns that fell into both, then you'd get into trouble when using either as a sanitizer if both were in scope (say from some barrier guard library).

Reply:
That's a good point. I hadn't previously considered it, and it hasn't caused us any problems, but we should probably call it out in our API docs as well. Thanks for flagging it up.